### PR TITLE
Nvprims native batch norm patch

### LINKED ISSUE #2088: codegen missing fp16 math support
### 🐛 Describe the bug

For the issues I'm seeing in our benchmark, I think we can work around it by making batch_norm promotion explicit in the graph. So I'll try to work around it in nvprim/primtorch.

Note: fp16 math can be explicitly put in our python API. Currently we don't have codegen support for these and it's going to be a real problem.

Repro script and error message below.

```
import torch
from torch._C._nvfuser import FusionDefinition, Fusion, DataType

def nvfuser_fusion(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1, -1, -1, -1], contiguous=[True, True, True, True], dtype=DataType.Half)
    T1 = fd.define_tensor(symbolic_sizes=[-1], contiguous=[True], dtype=DataType.Float)
    T2 = fd.define_tensor(symbolic_sizes=[-1], contiguous=[True], dtype=DataType.Float)
    T3 = fd.define_tensor(symbolic_sizes=[-1], contiguous=[True], dtype=DataType.Float)
    T4 = fd.define_tensor(symbolic_sizes=[-1], contiguous=[True], dtype=DataType.Float)
    S5 = fd.define_constant(0.100000)
    S6 = fd.define_constant(1.00000e-05)
    #T7, T8, T9 = fd.ops.batch_norm(T0, T1, T2, T3, T4, S5, S6)  # note: @kevinstephano python definition gives me an invalid script, which doesn't contain the scalar input to the op here.
    T7, T8, T9 = fd.ops.batch_norm(T0, T1, T2, T3, T4, True, S5, S6, False)
    T10 = fd.ops.cast(T7, dtype=DataType.Float)
    T11 = fd.ops.neg(T10)
    T12 = fd.ops.exp(T11)
    S13 = fd.define_constant(1.00000)
    T14 = fd.ops.add(S13, T12)
    S15 = fd.define_constant(1.00000)
    T16 = fd.ops.div(S15, T14)
    T17 = fd.ops.mul(T10, T16)
    T18 = fd.ops.cast(T17, dtype=DataType.Half)
    fd.add_output(T7)
    fd.add_output(T8)
    fd.add_output(T9)
    fd.add_output(T18)

inputs = [
    torch.randn(2, 3, 8, 8, device='cuda').half(),
    torch.randn(3, device='cuda'),
    torch.randn(3, device='cuda'),
    torch.randn(3, device='cuda'),
    torch.randn(3, device='cuda'),
]

fs = Fusion()
with FusionDefinition(fs) as fd:
    nvfuser_fusion(fd)

for _ in range(5) :
    fs.execute(inputs)
```

```


CUDA NVRTC compile error: __tmp_kernel1.cu(4488): error: no operator "/" matches these operands
            operand types are: CudaCodeGen::__half / CudaCodeGen::__half
          detected during instantiation of "void CudaCodeGen::welfordCombine(T &, T &, TN &, T, T, TN) [with T=CudaCodeGen::__half, TN=int]"
(7984): here

__tmp_kernel1.cu(4489): error: no operator "-" matches these operands
            operand types are: const CudaCodeGen::__half - CudaCodeGen::__half
          detected during instantiation of "void CudaCodeGen::welfordCombine(T &, T &, TN &, T, T, TN) [with T=CudaCodeGen::__half, TN=int]"
(7984): here

__tmp_kernel1.cu(4490): error: no operator "*" matches these operands
            operand types are: CudaCodeGen::__half * CudaCodeGen::__half
          detected during instantiation of "void CudaCodeGen::welfordCombine(T &, T &, TN &, T, T, TN) [with T=CudaCodeGen::__half, TN=int]"
(7984): here

__tmp_kernel1.cu(4491): error: no operator "*" matches these operands
            operand types are: CudaCodeGen::__half * CudaCodeGen::__half
          detected during instantiation of "void CudaCodeGen::welfordCombine(T &, T &, TN &, T, T, TN) [with T=CudaCodeGen::__half, TN=int]"
(7984): here

```

### Versions

`torchbenchPerf`

commit 124fba69b1101cece6d6d0b781c32a11e481dbc7 (HEAD, csarofeen/torchbenchPerf)
Author: Ivan Yashchuk <ivan.yashchuk@aalto.fi>
Date:   Mon Oct 17 16:34:38 2022 +0300

    Revert "Intercept aten._reshape_alias for nvFuser"
    
    This reverts commit eb24dbad03791483fe7a9f7906186dceec6aa6a2.

### PR DESCRIPTION
**Pull request by jjsjann123**
Sorta Fixes #2088 but not really.

Added explicit cast on inputs to `nvprims.native_batch_norm`. This avoids the explicit cast, which gives us issue on fusion definition.

Unfortunately I can't make a repro / fix without torchdynamo. Maybe @IvanYashchuk would be able to enlighten me. :laughing: 

Here's the python repro with dynamo that is fixed in this PR:
```
import torch

import torchdynamo
import functorch
from torchdynamo.optimizations.training import aot_nvprims_nvfuser
from torch._prims.context import TorchRefsNvfuserCapabilityMode
optimize = torchdynamo.optimize(aot_nvprims_nvfuser)

class Fusion(torch.nn.Module):
  def __init__(self) :
    super(Fusion, self).__init__()
    self.conv = torch.nn.Conv2d(32, 32, (1, 1), bias=False)
    self.bn = torch.nn.BatchNorm2d(32)

  def forward(self, inp) :
    out = self.conv(inp)
    out = out.relu()
    out = self.bn(out)
    return out

model = Fusion().cuda()
optimized_model = optimize(model)

input1 = torch.randn(2, 32, 8, 8, device="cuda")

with torch.cuda.amp.autocast():
  for _ in range(5):
    out = optimized_model(input1)
    out.sum().backward()
```

### CODE REVIEW

**Comment by jjsjann123** [comment]
@rdspring1 In the python script that I pasted here, if you turn off `autocast`, you can run the script without this BN fix and you'll notice the argument mismatch for `nvprims::full()`, which blocks the lowering to nvprims.

File: torch/_prims/nvfuser_prims.py [hunk 1]
```diff
@@ -445,12 +448,54 @@ def _prim_impl(
...
 
     nvprim_impl.impl(name, _prim_impl)
-    nvprim_autograd_impl.impl(
-        name, backwards_not_supported(torch.ops.nvprims.native_batch_norm.default)
-    )
-
     prim_packet = torch.ops.nvprims.native_batch_norm
     prim = prim_packet.default
+
+    def _native_batch_norm_ref(
```
Comment (IvanYashchuk) [inline torch/_prims/nvfuser_prims.py:454]: What is the output type of `mean` and `rstd`?
Would this work?
```suggestion
    @elementwise_type_promotion_wrapper(
        type_promoting_args=("input",),
        type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.COMPLEX_TO_FLOAT,
    )
    def _native_batch_norm_ref(
```

Comment (jjsjann123) [reply torch/_prims/nvfuser_prims.py:454]: type-promotion wrapper promotes all input and output tensors? So with elementwise promotion, output would also be in promoted math type, which is not what we want.

On paper, the promotion here for BN should only promote `input` and not on other things. e.g. you can't really promote running_stats, because those buffers should be in-place updated. weight/bias as well as mean/rstd should also come in `compute_math`.

Hence I'm only promoting things manually on `input` and cast `output` back to original dtype. BN behavior is strange....

Comment (IvanYashchuk) [reply torch/_prims/nvfuser_prims.py:454]: `type_promoting_args=("input",),` specifies that only `input` should be promoted.

Comment (jjsjann123) [reply torch/_prims/nvfuser_prims.py:454]: Lol, didn't read that carefully enough.
Does it cast outputs back to the original input dtype? I thought that's only available with reduction promotion.

Comment (IvanYashchuk) [reply torch/_prims/nvfuser_prims.py:454]: > Does it cast outputs back to the original input type?

Yes, but it might also try to cast all outputs, that's why I'm asking what's the expected type of `mean` and `rstd` outputs.

Comment (rdspring1) [reply torch/_prims/nvfuser_prims.py:454]: For "cpu", `mean` and `rstd` are cast back to input dtype.
https://github.com/pytorch/pytorch/pull/87116/files#diff-0e8654681a3dcc5fe5ed434f4fc23d2a47d69421eb261371aebc8754d8de5e06R2895-R2898

Comment (jjsjann123) [reply torch/_prims/nvfuser_prims.py:454]: > > Does it cast outputs back to the original input type?
> 
> Yes, but it might also try to cast all outputs, that's why I'm asking what's the expected type of `mean` and `rstd` outputs.

Got it.

Looks like cuda kernel just output mean/var as fp32.
```
In [24]: w = torch.randn(5, device="cuda")

In [25]: b = torch.randn(5, device="cuda")

In [26]: o = torch.ops.aten.native_batch_norm(x, w, b, None, None, True, 0.1, 1e-5)

In [27]: o[0].dtype
Out[27]: torch.float16

In [28]: o[1].dtype
Out[28]: torch.float32

In [29]: o[2].dtype
Out[29]: torch.float32
```

Comment (jjsjann123) [reply torch/_prims/nvfuser_prims.py:454]: Yep, looks like it indeed cast everything back to `input`. so we can't use it here if we want to preserve the behavior from aten.

```
In [33]: @elementwise_type_promotion_wrapper(
    ...:     type_promoting_args=("input",),
    ...:     type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.COMPLEX_TO_FLOAT,
    ...: )
    ...: def func(input, x):
    ...:     print(input.dtype)
    ...:     return (input + x, input, x)
    ...: 

In [34]: o = func(x, w)
torch.float32

In [35]: o[0].dtype
Out[35]: torch.float16

In [36]: o[1].dtype
Out[36]: torch.float16
```

Comment (jjsjann123) [reply torch/_prims/nvfuser_prims.py:454]: tagging @IvanYashchuk to get the approval. I'm leaning towards keeping the code as-is, until later we go fix all the batchnorm consistency issue between cpu/gpu. :vomiting_face:

**Review by IvanYashchuk — APPROVED** [review]
Merge it!
